### PR TITLE
Add PV scrub scripts to base Origin image

### LINKED ISSUE #3145: Determine image to use for Persistent Volume Recycling
The scrubber pod needs to use an approved and supported OpenShift image, not busybox.

The current implementation is this:

```
			Containers: []api.Container{
				{
					Name:  "scrubber",
					Image: "busybox",
					// delete the contents of the volume, but not the directory itself
					Command: []string{"/bin/sh"},
					// the scrubber:
					//		1. validates the /scrub directory exists
					// 		2. creates a text file in the directory to be scrubbed
					//		3. performs rm -rf on the directory
					//		4. tests to see if the directory is empty
					// the pod fails if the error code is returned
					Args: []string{"-c", "test -e /scrub && echo $(date) > /scrub/trash.txt && rm -rf /scrub/* && test -z \"$(ls -A /scrub)\" || exit 1"},
					VolumeMounts: []api.VolumeMount{
						{
							Name:      "vol",
							MountPath: "/scrub",
						},
					},
				},
			},
```

@smarterclayton

### PR DESCRIPTION
**Pull request by markturansky**
Resolves #3145.

Need guidance as to how this image name gets into NFS/HostPath volume plugins.

@smarterclayton

### CODE REVIEW

File: images/base/persistent_volume_scrubber.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Comment (ironcladlou) [inline images/base/persistent_volume_scrubber.sh]: Consider `-euo` for strictness?

File: images/base/persistent_volume_scrubber.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    # first and only arg is the directory to scrub
+    dir=$1
+
+    if [ -z "$dir" ]; then
+        echo "Usage:  scrub_directory some/path/to/scrub"
+        return 1
+    fi
+
+    if test ! -e $dir; then
```
Comment (ironcladlou) [inline images/base/persistent_volume_scrubber.sh]: Style looks odd to me (but I'm no bash guru)... any advantage over the more common `if [ ! -d "${dir}" ]`?

Comment (markturansky) [inline images/base/persistent_volume_scrubber.sh]: Meant to replace all the inline ```test```s I had with ```if``` statements.

-e is "exists", which is sufficient to determine if the directory is there.

File: images/base/Dockerfile [hunk 3]
```diff
@@ -9,3 +9,7 @@ FROM centos:centos7
 # components from EPEL must be installed in a separate yum install step
 RUN yum install -y git tar wget socat hostname sysvinit-tools util-linux epel-release && \
     yum clean all
+
+ADD persistent_volume_scrubber /usr/local/bin/
```
Comment (smarterclayton) [inline images/base/Dockerfile]: You should create a script directory and then add the entire script directory.  The script directory should be under the root `scripts/volumes` and the filename should be `recycle.sh`.  You should add `scripts` into `/usr/share/openshift` or whatever the appropriate path location is for installing client specific scripts.  You shouldn't need to chmod if the file in source control is properly chmodded.

**Comment by smarterclayton** [comment]
The recycler needs to take an argument for the image name to start, the value of the image name argument should be generated like the other components.  For now, I would reuse the deployer image (so your pod will be `image.For("deployer")` and your executable should be `/usr/share/openshift/scripts/volume/recycle.sh`

File: images/base/persistent_volume_scrubber.sh (hunk 2 above)
Comment (smarterclayton) [inline images/base/persistent_volume_scrubber.sh]: We should always be using [ ! ]

File: images/base/persistent_volume_scrubber.sh (hunk 1 above)
Comment (smarterclayton) [inline images/base/persistent_volume_scrubber.sh]: Follow the pattern in our hack scripts - three explicit lines at the top, below the description of the script.

**Comment by markturansky** [comment]
To have configurable images for the scrubber, wouldn't we need all three:  image name, command, and args?  I imagine I need a struct and a map of these structs by plugin name.  

I think VolumeHost is meant as the interface to the world for a plugin.  The plugin could get their config by name and use that when creating their Recycler object.

This is going to require a patch upstream and quick design review so that I can keep it in sync downstream.

**Comment by markturansky** [comment]
Put another way, I can make upstream generically configurable per above, but then I can hard-code our image where I set up the recycler controller.  Upstream can continue to use the default implementation without override.

**Comment by markturansky** [comment]
Per our email conversation, perhaps there is an easier way that I am missing or a short term solution that can bridge the gap until there's a correct.

It seems like the correct solution for configuring recyclers should use VolumeHost.  Each plugin provides a default implementation for scrubbing and allows override (via the Host) through whatever configuration mechanism available (which affects the Host).  Plugins get their config from the Host.

Anything else results in many leaks.  For example, I can make a map of configs and pass it to the Recycler controller, but that now requires knowing the name of every volume plugin, which is unknown to all except the binary (the constant value names are even package protected variables).

Is there a bandaid to apply in the near term?  I fear this request for a configurable scrubber image is too big to fit into a last minute patch.

**Comment by smarterclayton** [comment]
If you don't think you can safely wire the values for the recycler image template down into nfsRecycler today and tomorrow, we'll just have to disable recycling for 1.0.0 and fix it afterwards.

However, it seems like this is something you can pass through to ProbeVolumePlugins pretty late in the binding.  

----- Original Message -----

**Comment by markturansky** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/pull/9870

This will allow us to override the recycler, even though we will need to hard-code the plugin names in the binary.

**Comment by smarterclayton** [comment]
I'm not sure why you aren't configuring the plugin, not the recycler.  ProbeVolumePlugins can take arguments... just like anything else.

**Comment by smarterclayton** [comment]
This is fine as an upstream patch though for now.

**Comment by markturansky** [comment]
Is one way preferable over the other?

Args to the plugins would require OS to specify its own plugins.  Today it gets them from the upstream binary.

**Comment by smarterclayton** [comment]
Recycler implementation seems much more coupled to server config than the volume host is.  VolumeHost is more about dynamic things (which pods are running) instead of static immutable config.  Generally, we pass static immutable config into our functions.

Maintaining a list of plugins in our code is not a serious issue.  In fact, it's generally preferable.

----- Original Message -----

**Comment by markturansky** [comment]
Upstream PR was updated with your feedback.  I'm going to patch it into OS now and make it work.

**Comment by markturansky** [comment]
Upstream PR is merged into OS.  Hopefully it looks close to good enough.

I'm using the deployer image as the base but kept (for now) the same scriptlet in the command of the pod.  I'll work on our own image next, but I think the pressing need to remove busybox is alleviated.

**Comment by smarterclayton** [comment]
You need to separate the upstream part

**Comment by markturansky** [comment]
@smarterclayton the upstream piece is now isolated in its own commit.  do you need a separate PR for that or is a separate commit enough?

The recycle.sh script (with chmod +x) is in the origin image and I changed the recycler pods to use it.

**Comment by markturansky** [comment]
@smarterclayton nm on the upstream thing.  spacing out this morning.  already submitted that PR.  will remove that from this one.

**Comment by markturansky** [comment]
@smarterclayton I removed the upstream patch from this PR.  Please let me know if there is anything left to do with this PR for now.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Comment by markturansky** [comment]
@smarterclayton This PR will fail until #3263 gets merged.

**Comment by markturansky** [comment]
i've tested through the images locally and all is well.  I think we're just fighting a Jenkins battle now to get this in before the deadline.

**Comment by smarterclayton** [comment]
[test]

**Comment by markturansky** [comment]
@smarterclayton this needs to get kicked off again to pull in the upstream changes that just got merged through the other PR

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**